drud/ddev
# Trust SSL/TLS Certificates, fixes #1247, fixes #1501, fixes #849

## The Problem/Issue/Bug:

https is now standard everywhere, but ddev has never had a way to present a trusted certificate.
mkcert makes it possible.

## How this PR Solves The Problem:

* Use mkcert to create a trusted CA on the host
* Copy the trusted CA files into the router and the webserver container
* Generate certs for each name needed (or wildcard)
* Provide a cert for 127.0.0.1 for web container
* Make sure web container trusts the certs as well
* Reorganize nginx config so that there's not so much repetition
* Add simple include nginx config add-ons per #1501 - simple item while doing the nginx config reorg

## Manual Testing Instructions:

* Install mkcert (`brew install mkcert`, `choco install -y mkcert`)
* `mkcert -install`
* Use browser and curl against https URLs on host
* Test complex URLs introduced with additional_hostnames and additional_fqdns. 
* Try just regular old https
* Try curl within the web container against localhost and against the FQDN with https
* Try curl on the host against localhost-https port as shown in `ddev describe`
* Test behavior without mkcert installed. It should "work", but not have trusted browser responses.
* Use nginx snippet overrides in `.ddev/nginx/<something>.conf` and verify that they get detected.
* Review updated docs for mkcert and nginx config and other (docs/index.md, docs/users/extend/customization-extendibility.md
* Test `ddev config --host-https-port`

## Automated Testing Overview:

- [x] We may have places where we're currently ignoring https failure and we can respect it now!
- [x]  Start a project with no router running
- [x]  Start up with another project running, both should be valid
- [x]  Container tests for ddev-router, cert should work inside (`curl https://d8composer.ddev.local --resolve "d8composer.ddev.local:443:127.0.0.1"`)
- [x]  Container tests for ddev-webserver, cert should work inside against 127.0.0.1 address. #1235 will prevent access to router via hostname

## Related Issue Link(s):

#1247 

## Release/Deployment notes:

* Add mkcert dependency to homebrew
* Explain in release how nginx snippets work
* Explain mkcert in release
* Explain `ddev config --host-https-port`
* Explain that https is now the first option in lists of URLs and why
